File: 22-merge-intervals/solution.cpp
// Merge Overlapping Intervals
// Time: O(n log n), Space: O(n)

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if (intervals.empty()) return {};
        
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> merged;
        merged.push_back(intervals[0]);
        
        for (int i = 1; i < intervals.size(); i++) {
            vector<int>& current = intervals[i];
            vector<int>& last = merged.back();
            
            if (current[0] <= last[1]) {
                last[1] = max(last[1], current[1]);
            } else {
                merged.push_back(current);
            }
        }
        
        return merged;
    }
};

// Test
int main() {
    Solution sol;
    vector<vector<int>> intervals1 = {{1, 3}, {2, 6}, {8, 10}, {15, 18}};
    vector<vector<int>> result1 = sol.merge(intervals1);
    
    for (auto& interval : result1) {
        cout << "[" << interval[0] << ", " << interval[1] << "] ";
    }
    cout << endl;  // [1, 6] [8, 10] [15, 18]
    
    vector<vector<int>> intervals2 = {{1, 4}, {4, 5}};
    vector<vector<int>> result2 = sol.merge(intervals2);
    
    for (auto& interval : result2) {
        cout << "[" << interval[0] << ", " << interval[1] << "] ";
    }
    cout << endl;  // [1, 5]
    
    return 0;
}

File: 22-merge-intervals/solution.go
// Merge Overlapping Intervals
// Time: O(n log n), Space: O(n)

package main

import (
	"fmt"
	"sort"
)

type Solution struct{}

func (s *Solution) merge(intervals [][]int) [][]int {
	if len(intervals) == 0 {
		return [][]int{}
	}
	
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})
	
	merged := [][]int{intervals[0]}
	
	for i := 1; i < len(intervals); i++ {
		current := intervals[i]
		last := merged[len(merged)-1]
		
		if current[0] <= last[1] {
			if current[1] > last[1] {
				last[1] = current[1]
			}
		} else {
			merged = append(merged, current)
		}
	}
	
	return merged
}

// Test
func main() {
	sol := &Solution{}
	result1 := sol.merge([][]int{{1, 3}, {2, 6}, {8, 10}, {15, 18}})
	fmt.Println(result1) // [[1 6] [8 10] [15 18]]
	
	result2 := sol.merge([][]int{{1, 4}, {4, 5}})
	fmt.Println(result2) // [[1 5]]
}
